### PR TITLE
Fix/slow voices

### LINKED ISSUE #670: [Bug] slow PCM voices since commit 252f535ebd82aa1cc311a1aa002c7c0875b2f251
Investigating...

This is tested with Dune.

### PR DESCRIPTION
**Pull request by maximilien-noal**
fixed: slow PCM voices
    
    Voices are normal each time I test them now.
    
    To make rendering faster, I rolled the sample conversion, stereo effect
    conversion, and volume conversion into one single loop in the sofware
    mixer.

fixes #670

### CODE REVIEW

File: src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs [hunk 1]
```diff
@@ -24,54 +23,59 @@ public sealed class SoftwareMixer : IDisposable, IDebuggableComponent {
...
         }
-        ApplyStereoSeparation(data, channel);
-        ApplyVolume(data, channel);
-        return _channels[channel].WriteData(data);
+        float volumeFactor = channel.Volume / 100f;
+        float separation = channel.StereoSeparation / 100f;
+
+        Span<float> target = stackalloc float[data.Length];
+        for (int i = 0; i < data.Length; i++) {
+            target[i] = (data[i] / 32768f)  * volumeFactor * (1 - separation);
```
Comment (JorisVanEijden) [inline src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs:60]: no need to calculate `volumeFactor * (1 - separation)` in each loop iteration. I'd factor this out into a variable before entering the loop

File: src/Spice86.Core/Backend/Audio/AudioPlayer.cs [hunk 2]
```diff
@@ -56,10 +43,10 @@ public abstract class AudioPlayer : IDisposable
     /// Writes the audio data to the rendering backend
     /// </summary>
     /// <param name="data">The data to write</param>
-    /// <returns>The length of data written. Might not be equal to the input data length.</returns>
-    internal abstract int WriteDataInternal(Span<float> data);
+    /// <returns>The length of data written. Equals to the input data length.</returns>
```
Comment (JorisVanEijden) [inline src/Spice86.Core/Backend/Audio/AudioPlayer.cs:46]:
```suggestion
    /// <returns>The length of data written. Equal to the input data length.</returns>
```

File: src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs [hunk 3]
```diff
@@ -24,54 +23,59 @@ public sealed class SoftwareMixer : IDisposable, IDebuggableComponent {
...
         }
-        ApplyStereoSeparation(data, channel);
-        ApplyVolume(data, channel);
-        return _channels[channel].WriteData(data);
+        float volumeFactor = channel.Volume / 100f;
+        float separation = channel.StereoSeparation / 100f;
+        
+        Span<float> target = stackalloc float[data.Length];
+        for (int i = 0; i < data.Length; i++) {
+            target[i] = ((data[i] - 127) / 128f)  * volumeFactor * (1 - separation);
```
Comment (JorisVanEijden) [inline src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs:76]: Same thing here, move the calculation of the multiplier outside the loop

File: src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs (hunk 1 above)
Comment (maximilien-noal) [reply src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs]: Done.

File: src/Spice86.Core/Backend/Audio/AudioPlayer.cs (hunk 2 above)
Comment (maximilien-noal) [reply src/Spice86.Core/Backend/Audio/AudioPlayer.cs:46]: Applied.

File: src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs (hunk 3 above)
Comment (maximilien-noal) [reply src/Spice86.Core/Emulator/Devices/Sound/SoftwareMixer.cs]: Done.

**Review by JorisVanEijden — APPROVED** [review]

**Pull request merged**